osu.Game/Rulesets/Mods/ModChallenge.cs

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Linq;
+using osu.Framework.Bindables;
+using osu.Game.Configuration;
+using osu.Game.Rulesets.Objects;
+using System.Collections.Generic;
+using osu.Game.Beatmaps;
+using osu.Game.Beatmaps.Timing;
+using osu.Game.Scoring;
+using osu.Game.Rulesets.Judgements;
+using osu.Game.Rulesets.Scoring;
+using System.ComponentModel;
+
+namespace osu.Game.Rulesets.Mods
+{
+    public abstract class ModChallenge : ModFailCondition, IApplicableToBeatmap
+    {
+        public override ModType Type => ModType.DifficultyIncrease;
+        public override double ScoreMultiplier => 1;
+        public override bool RequiresConfiguration => true;
+        public override Type[] IncompatibleMods => base.IncompatibleMods.Append(typeof(ModPerfect)).Append(typeof(ModEasyWithExtraLives)).ToArray();
+
+        [SettingSource("Checking interval", "When conditions will be checked.")]
+        public Bindable<ChallengeCheckInterval> CheckingInterval { get; } = new Bindable<ChallengeCheckInterval>();
```

Comment:
I think this setting should be excluded from the initial implementation. It's difficult to explain to the user and understand, and it's not clear to me that anyone is ever going to use anything but "continuously". Its presence also overcomplicates the implementation.

Reply:
For the case of "at end" at least, this would be useful for preventing submission if the play isn't good enough, but still being able to play through the map (something people have been asking for)

Reply:
A lot of that is a byproduct of how stable submission works and the threat of getting "score v1'd". I don't think that argument is valid at least for scoreboard purposes anymore, with the reasoning being, that (a) if we sort out leaderboards to remove getting "score v1'd", then there may no longer be any need to prevent submission, and (b) if you wanted to add a "pp challenge" mod, then the *only* time it'd make sense to check pp at is at the end of the map.

Reply:
Anyway I'm not saying that the setting should *never* be there, just that the scope should be reduced *for now*. As stated in my other comment I'm not hugely against it, but this pull has so much stuff jammed into it that it's really hard to work out the details of everything (which are dodgy here) and I'm kind of forced to accept or deny this at face value as-is.